index.ts

```diff
@@ -655,135 +414,122 @@ class Rabbitr extends EventEmitter {
               }
             }
 
-            return response;
-          }, error => {
-            // TODO - investigate why bluebird wraps the error here with an object
-            if (error && error.isOperational && error.cause) {
-              throw error.cause;
-            }
-            throw error;
+            completed(response);
           }
-        ).catch(TimeoutError, error => {
-          this.log(`request timeout firing for ${rpcQueue} to ${returnQueueName}`);
-          throw error;
-        });
+          catch(err) {
+            // send the failure to the rejection of the promise
+            failed(err);
+          }
+        };
+
+        try {
+          await channel.consume(replyQueue, gotReply, {noAck: true});
+        }
+        catch(err) {
+          failed(err);
+        }
       });
-    }).asCallback(cb);
-  }
 
-  rpcListener(topic: string, executor: Rabbitr.IRpcListenerExecutor<any, any>, callback?: Rabbitr.ErrorCallback): Bluebird<void>;
-  rpcListener(topic: string, opts: Rabbitr.IRpcListenerOptions<any, any>, executor: Rabbitr.IRpcListenerExecutor<any, any>, callback?: Rabbitr.ErrorCallback): Bluebird<void>;
-  rpcListener<TInput, TOutput>(topic: string, executor: Rabbitr.IRpcListenerExecutor<TInput, TOutput>, callback?: Rabbitr.ErrorCallback): Bluebird<void>;
-  rpcListener<TInput, TOutput>(topic: string, opts: Rabbitr.IRpcListenerOptions<TInput, TOutput>, executor: Rabbitr.IRpcListenerExecutor<TInput, TOutput>, callback?: Rabbitr.ErrorCallback): Bluebird<void>;
+      // send the request now
+      const request = {
+        d: data,
+        returnQueue: replyQueue,
+        expiration: now + timeoutMS,
+      };
+      this.log('sending rpc request');
+      this._publishChannel.sendToQueue(this._formatName(rpcQueue), new Buffer(stringify(request)), {
+        contentType: 'application/json',
+        expiration: `${timeoutMS}`,
+      });
 
-  rpcListener<TInput, TOutput>(topic: string, opts: Rabbitr.IRpcListenerOptions<TInput, TOutput>, executor?, callback?: Rabbitr.ErrorCallback): Bluebird<void> {
-    // istanbul ignore next
-    if (!this.connectionPromise.isFulfilled()) {
-      // delay until ready
-      return this.whenReady().then(() =>
-        this.rpcListener(topic, opts, executor)
+      // race the result against a timeout promise
+      const result = await racePromises<TOutput>(
```

Comment:
1. Won't it throw even if `resultPromise` resolves first?  ? 2. Can't you use Promise.race ?